src/cogent3/core/new_moltype.py

```diff
@@ -529,30 +529,143 @@ def _(self, seq: numpy.ndarray) -> bool:
         return (seq == self.degen_gapped_alphabet.gap_index).any()
 
     @functools.singledispatchmethod
-    def get_degenerate_positions(self, seq, include_gap=True) -> numpy.ndarray:
-        """returns a boolean array indicating degenerate positions"""
+    def get_degenerate_positions(self, seq, include_gap=True) -> list[int]:
+        """returns indices matching degenerate characters"""
         raise TypeError(f"{type(seq)} not supported")
 
     @get_degenerate_positions.register
-    def _(self, seq: numpy.ndarray, include_gap=True) -> numpy.ndarray:
+    def _(self, seq: numpy.ndarray, include_gap=True) -> list[int]:
 
         for index, val in enumerate(self.degen_gapped_alphabet):
             if include_gap and val in self.gap or val in self.ambiguities:
                 break
-        return seq >= index
+        degens = seq >= index
+        return numpy.where(degens)[0].tolist()
 
     @get_degenerate_positions.register
-    def _(self, seq: str, include_gap=True) -> numpy.ndarray:
+    def _(self, seq: str, include_gap=True) -> list[int]:
         return self.get_degenerate_positions(
             self.degen_gapped_alphabet.to_indices(seq), include_gap
         )
 
     @get_degenerate_positions.register
-    def _(self, seq: bytes, include_gap=True) -> numpy.ndarray:
+    def _(self, seq: bytes, include_gap=True) -> list[int]:
         return self.get_degenerate_positions(
             self.degen_gapped_alphabet.to_indices(seq), include_gap
         )
 
+    @functools.singledispatchmethod
+    def degap(self, seq) -> StrORBytesORArray:
+        """removes all gap and missing characters from a sequence"""
+        # refactor: design
+        # previous implementation also removed missing characters, as does this
+        # implementation. But should we revisit this?
+        raise TypeError(f"{type(seq)} not supported")
+
+    @degap.register
+    def _(self, seq: str) -> str:
+        trans = dict([(i, None) for i in map(ord, self.gaps)])
```

Comment:
add a comment to refactor, we should be caching translation callables rather than doing this on every method call. We should make bytes the primary method and an instance of the convert_alphabet which deletes the gap character.

Reply:
@GavinHuttley, I will add a comment about caching. The reason I didn't send all dispatches through bytes is that I thought it would be best to return the same type as the original sequence.

Reply:
have added a comment about making bytes the primary method too